icefall/utils.py

```diff
@@ -411,3 +411,20 @@ def write_error_stats(
 
         print(f"{word}   {corr} {tot_errs} {ref_count} {hyp_count}", file=f)
     return float(tot_err_rate)
+
+
+def is_main_process():
+    """Check if the current process is main.
+
+    - For DistributedDataParallel (DDP) mode:
+        The current is main process if its rank is 0, and its rank get from
+        os.environ["RANK"] (which already be set by DDP mode).
```

Comment:
> os.environ["RANK"] (which already be set by DDP mode).

We are not using `torch.distributed.launch` to do DDP training and I am afraid
`RANK` is not an environment variable in our training script.

Could you use `torch.distributed.get_rank()` to get the rank of a node?

Example usages can be found in lhotse, e.g.,

https://github.com/lhotse-speech/lhotse/blob/master/lhotse/dataset/sampling/base.py#L77

```python
    def _maybe_init_distributed(self, world_size: Optional[int], rank: Optional[int]):
        if world_size is not None:
            assert world_size >= 1
        if rank is not None:
            assert rank >= 0
        if not dist.is_available() or not dist.is_initialized():
            self.world_size = 1 if world_size is None else world_size
            self.rank = 0 if rank is None else rank
            return
        self.world_size = dist.get_world_size() if world_size is None else world_size
        self.rank = dist.get_rank() if rank is None else rank
        assert self.rank < self.world_size
```

---

`torch.distributed.launch` is useful for multi-node multi-GPU training. Currently, icefall
supports only single-node multi-GPU training.